xrviz/dashboard.py

```diff
@@ -0,0 +1,93 @@
+import panel as pn
+import xarray as xr
+import hvplot.xarray
+from cartopy import crs as ccrs
+from .sigslot import SigSlot
+from .control import Control
+from .utils import convert_widget
+
+
+class Dashboard(SigSlot):
+    """
+    This section provides access to the complete generated dashboard,
+    consisting of all subsections.
+
+    Parameters
+    ----------
+    data: `xarray` instance: `DataSet` or `DataArray`
+           datset is used to initialize.
+
+    Attributes
+    ----------
+    panel: Displays the generated dashboard.
+    control: Provides access to the generated control panel.
+    index_selectors: Provides access to generated index_selectors.
+    plot: Plot button, upon click generates graph according to
+          kwargs selected in other sub-sections.
+    output: Provides access to generated graph.
+    """
+
+    def __init__(self, data):
+        super().__init__()
+        self.data = data
+        self.control = Control(self.data)
+        self.plot = pn.widgets.Button(name='Plot', width=200)
+        self.index_selectors = self.control.fields.index_selectors
+        self.output = pn.Row(pn.Spacer(name='Graph'))
+
+        self._register(self.plot, 'plot_clicked', 'clicks')
+        self.connect('plot_clicked', self.create_plot)
+        self.control.displayer.connect('variable_selected', self.check_is_plottable)
+
+        self.panel = pn.Column(self.control.panel,
+                               self.plot,
+                               self.output
+                               )
+        if isinstance(self.data, xr.Dataset):
+            self.plot.disabled = True
+        else:
+            if self.data.name in self.data.coords or len(self.data.dims) <= 1:
+                self.plot.disabled = True
+
+    def create_plot(self, *args):
+        kwargs = self.control.kwargs
+        var = kwargs['']
+        graph_opts = {}
+        graph_opts['x'] = kwargs['x']
+        graph_opts['y'] = kwargs['y']
+        graph_opts['rasterize'] = True
+        graph_opts['width'] = 600
+        graph_opts['height'] = 400
+        graph_opts['crs'] = ccrs.PlateCarree()
+
+        if isinstance(self.data, xr.Dataset):
+            graph_opts['title'] = var
+            self.graph = pn.Row(self.data[var].hvplot.quadmesh(**graph_opts))
+        else:
+            graph_opts['title'] = self.data.name
+            self.graph = pn.Row(self.data.hvplot.quadmesh(**graph_opts))
+
+        self.output[0] = self.graph[0][0]
```

Comment:
Why do we make a `Row`, above, and then extract it's components here?

Reply:
```
data = xv.sample_data.great_lakes
data = data.set_coords(['lat','lon'])
generated_plot  = data.temp.hvplot.quadmesh(x ='lat',y='lon')
print(generated_plot)
```
returns `:DynamicMap   [sigma,time]`. This means that we can't separate the sliders and plot. If we enclose the `generated_plot` in a `Row` we can easily extract the components.
```
graph = pn.Row(generated_plot)
print(graph)
```
returns
```
Row
    [0] Row
        [0] HoloViews(DynamicMap)
        [1] Column
            [0] DiscreteSlider(name='Sigma Stretched V..., options=OrderedDict([('0', ...]), value=0.0)
            [1] DiscreteSlider(name='Time', options=OrderedDict([('2019-03-20 ...]), value=numpy.datetime64('2019-03-...)
```
so `graph[0][0]` refers to the actual plot in the `generated_plot` object.

Reply:
Indeed, but is it not the case that if
```
self.graph = self.data.hvplot.quadmesh(**graph_opts)
```
then you only need
```
self.graph[0]
```
?

Reply:
I have tried this for dataArray using `ds.temp`. If we do `self.graph[0]` then the entire  generated output (including the default sliders are displayed). So `self.graph[0][0]` has  to be used.

Reply:
This has been changed now.